### ISSUE TITLE
New Error format in New Relic

### ISSUE THREAD

**Issue by netalico** [ProxiBlue/NewRelic#25]
About a week ago my errors in New Relic changed to be a lot longer so that they included "Noticed Exception...". Example below.

Any idea what caused this and how I can get them to shorten back to their original length? 

Mage_Core_Exception: Noticed exception 'Mage_Core_Exception' with message 'Please specify the product's option(s).' in /home/freeflyapparel/public_html/app/Mage.php:595

<img width="1205" alt="screen shot 2015-07-15 at 8 23 04 am" src="https://cloud.githubusercontent.com/assets/2094614/8699190/ba5b7212-2ad1-11e5-9147-4c7a2d2e51eb.png">

**Comment by ProxiBlue** [comment]
Did you update anything, especially to the latest version of this extension?
Can you expand on the error, an get a trace. I would like to see the code flow path that leads to the error.

Additionally, you can exclude those exceptions from error push by placing the text ''Please specify the product's option(s).' in the exclude list in admin. They are not really exceptions / errors

**Comment by netalico** [comment]
I definitely didn't update the extension. I noticed it changed on all my Magento sites across multiple servers so its kind of confounding. 

I have excludes for those kind of things, but they stopped working once it switched to this new formatting.

**Comment by netalico** [comment]
Here's the trace from NR:

…at /home/freeflyapparel/public_html/app/code/core/Mage/Checkout/Model/
Cart.php (284)
…apparel/public_html/app/code/core/Mage/Checkout/controllers/
CartController.php (218)
…reeflyapparel/public_html/app/code/core/Mage/Core/Controller/Varien/
Action.php (418)
…arel/public_html/app/code/core/Mage/Core/Controller/Varien/Router/
Standard.php (254)
…freeflyapparel/public_html/app/code/core/Mage/Core/Controller/Varien/
Front.php (172)
…lled at /home/freeflyapparel/public_html/app/code/core/Mage/Core/Model/
App.php (354)
…ge_Core_Model_App::run called at /home/freeflyapparel/public_html/app/
Mage.php (684)
in Mage::run called at /home/freeflyapparel/public_html/index.php (71)

**Comment by ProxiBlue** [comment]
So, a key thing here is of course the 'Notice Exception' part, which is not normal.

Checking my extension, I can confirm that there is no such string anywhere in it, so not added by it:

[22:50:50][lucas@dualvac ~/workspace/3rdparty/NewRelic](master)$ grep -r "Notice Exception" .
[22:50:59][lucas@dualvac ~/workspace/3rdparty/NewRelic](master)$ 

Checking all of magento install (1.9.01 CE)

[22:54:20][lucas@dualvac ~/workspace/personal/giftpromo](master $)$ grep -r "Notice Exception" .
[22:54:23][lucas@dualvac ~/workspace/personal/giftpromo](master $)$ 

Can you confirm the Exception file located in /app/code/local/Mage/Core is this one (there could be slight differences as the module namespace was changed to proxiblue_newrelic in the refered link:
https://github.com/ProxiBlue/NewRelic/tree/master/app/code/local/Mage/Core

Can you grep your magento folder (from the root) to see if the given text appears in any file

use :  grep -r "Notice Exception" .

**Comment by ProxiBlue** [comment]
Another thought. Possibly you have an older version of extension, that did contain that string. What version you got installed?

**Comment by netalico** [comment]
I had 0.1.11. I'll go ahead and update to the latest. Even if its not the root cause I figure it can't hurt.

**Comment by ProxiBlue** [comment]
Just grepped 0.1.11, also no occurance of just 'Noticed'

[23:10:35][lucas@dualvac ~/tmp/0.1.11]$ grep -r "Noticed" .
[23:10:39][lucas@dualvac ~/tmp/0.1.11]$

**Comment by netalico** [comment]
Did the grep and no mention of "Noticed" on my server either.

**Comment by ProxiBlue** [comment]
Stranger and stranger. Pondering. May send you an altered file to add in
some debug trace in the exceptions file. Basically to see what it is
getting and what it is sending.

About to crash. Nearly midnight here.

Is option to ignore debug messages disabled in config of module?
Is magento developer mode disabled on servers?

**Comment by netalico** [comment]
Here's my settings. Developer mode is disabled. Thanks for looking into this!

![screen shot 2015-07-15 at 11 30 46 am](https://cloud.githubusercontent.com/assets/2094614/8702287/faee91b6-2ae4-11e5-9a9e-991f0ac0f0d5.png)

**Comment by ProxiBlue** [comment]
Tomorrow, I'll run some tests on a site where I have the extension installed. see if  get the same.

**Comment by ProxiBlue** [comment]
Also, possible you can determine if the newrelic agent updated recently?

**Comment by netalico** [comment]
I'm not sure how to tell if it updated recently, but my version is 4.23.1.107

**Comment by ProxiBlue** [comment]
It will be noted in the agent log file on server. I know the agents on our
cpanel boxes self update, and restart themselves periodically. Right now i
cannot recall where that log is located. Possible in /var/log

**Comment by netalico** [comment]
Looks like its updated at least twice this month. It was 4.22.0 about 2 weeks ago.

**Comment by ProxiBlue** [comment]
Hi,

So firstly having had a look at historic log entries for pushed errors, I can 100% determine the 'Noticed exception' is paced by New Relic. 
As an example, from the docs, an image of a test error sent to New Relic:

![selection_009](https://cloud.githubusercontent.com/assets/4994260/8711368/e40b4d46-2b83-11e5-938c-ddab9810e305.png)

The New Relic command used, is called 'newrelic_notice_error', which fits.

Using the extension built in testing exception push, I pushed a test exception. The end result is as such:

![selection_010](https://cloud.githubusercontent.com/assets/4994260/8711401/2ab0c3d4-2b84-11e5-8948-dda2744cc512.png)

Can you do that, and lets see what your end result is.

The url to use is:

<your site url>/newrelic/test/exceptionlog/key/<YOUR API KEY>

It looks to me, like something has changed in New Relic/or at least your new relic settings possibly.

Also, you mentioned you have exclusions in your extenion to block these 'false positives', but in the image of your settings, this is not so. Is it simply possible the block lists were accidentally cleared out?

ALso the trace you sent me does not show the extensions code has been used at all. The trace shoudl at least end with teh file that sent the error (ie a proxiblue file) to newrelic. As an example, notice the last call in the trace below:

Note the last entry is the file that sent the error:

Stack trace
…/new_html/app/code/community/ProxiBlue/NewRelic/controllers/
TestController.php (37)
…at /home/enjocom/new_html/app/code/core/Mage/Core/Controller/Varien/
Action.php (419)
…enjocom/new_html/app/code/core/Mage/Core/Controller/Varien/Router/
Standard.php (254)
… at /home/enjocom/new_html/app/code/core/Mage/Core/Controller/Varien/
Front.php (176)
…ispatch called at /home/enjocom/new_html/app/code/core/Mage/Core/Model/
App.php (354)
in Mage_Core_Model_App::run called at /home/enjocom/new_html/app/Mage.php (683)
in Mage::run called at /home/enjocom/new_html/index.php (86)

I don't actually have any real errors to check (sites are 100% error free ;) ) - I will cause a real one today, just to confirm what I say is true  ;) 

SO, just to confirm, the issue here is that you are no longer blocking the messages, or just that they have changed in syntax?

**Comment by ProxiBlue** [comment]
I am starting to suspect that the error you see in New Relic is not actually sent by the ProxiBlue extension. It is an error trapped by the newrelic agent, directly from PHP...this is why my code doe snot show in the trace. WIll kknow better once we see the result of the test exception, and its output.

What is the value of the newrelic setting:

newrelic.loglevel (ref: https://docs.newrelic.com/docs/agents/php-agent/configuration/php-agent-configuration#inivar-settings)

**Comment by netalico** [comment]
I agree that it seems to be the NR PHP Agent thats adding the additional verbiage. My log level is set to info. I'll do the other test you recommended.

**Comment by netalico** [comment]
Here's my trace.

Error message
Mage_Core_Exception: Noticed exception 'Mage_Core_Exception' with message 'Test exception thrown' in /home/freeflyapparel/public_html/app/Mage.php:595

Stack trace
…blic_html/app/code/community/ProxiBlue/NewRelic/controllers/
TestController.php (37)
…reeflyapparel/public_html/app/code/core/Mage/Core/Controller/Varien/
Action.php (418)
…arel/public_html/app/code/core/Mage/Core/Controller/Varien/Router/
Standard.php (254)
…freeflyapparel/public_html/app/code/core/Mage/Core/Controller/Varien/
Front.php (172)
…lled at /home/freeflyapparel/public_html/app/code/core/Mage/Core/Model/
App.php (354)
…ge_Core_Model_App::run called at /home/freeflyapparel/public_html/app/
Mage.php (684)
in Mage::run called at /home/freeflyapparel/public_html/index.php (71)

**Comment by ProxiBlue** [comment]
Interesting.

As you can see, and the reason for this query, is that in my example, the message is purely the actual exception message 'test exception throw' where yours is the longer version.

If you look at the code for this test script: https://github.com/ProxiBlue/NewRelic/blob/master/app/code/community/ProxiBlue/NewRelic/controllers/TestController.php you can see it simply throws an exception via mage code:

Mage::throwException('Test exception thrown');

so, we are back to needing to see what the exception message is that my custom exception logger is pushing.

Basically, it gets the simple message, and pushes that same simple message, the issue is something that changed in regards to your NR setup.

The site I am testing on is using a very old agent: 4.11.0.63, so that can most definitely be the resulting factor! I will get a more modern agent test done.

**Comment by ProxiBlue** [comment]
Note: In latest version the test url changed to:

proxiblue_newrelic/test/exceptionlog/key/

was previously 

newrelic/test/exceptionlog/key/

**Comment by ProxiBlue** [comment]
Ok, matched.

Using the latest version of teh extension, with a newer version of the agent: I get teh same exception message as you:

![selection_176](https://cloud.githubusercontent.com/assets/4994260/8716117/28f42de8-2bbf-11e5-8a3d-e7c8149c3e55.png)

Since you did not update to the latest version of the extension, I can say with confidence that the change in the message is due to the newer version of the newrelic agent (in this test I used 4.20.2.95) 

SO, reason found!

Ok, so what is not working? or is the query only that the message got larger?

**Comment by ProxiBlue** [comment]
AH, but hold on, I need to test further. Is it possible that the later newrelic extension is bypassing my code, and is reporting the original thrown exception directly?

**Comment by ProxiBlue** [comment]
Ok, so tested further, and I can confirm 100% that I am simply pushing teh text exception string to NR using the given newrelic command of 'newrelic_notice_error'

The issue is thus with newrelic, and the newer version of the agent, which seems cause the longer message to appear in the UI. It may be possible the agent is adjusting the string, wrapping it in additional data.

See image below, taken from PhpStorm, in a debug session just before the data is pushe dto newrelic. This clearly shows I am simply pushing the simple string, not that entire message.

![selection_178](https://cloud.githubusercontent.com/assets/4994260/8716236/1b6e8978-2bc1-11e5-94ed-5135de873551.png)

Unfortunately nothing I can do about this part, but if I can understand your related issue caused by the longer message, I can try and help with that, if possible.

**Comment by netalico** [comment]
I gotcha. I think the real remaining "issue" is that I'm getting occasional empty message exceptions, like the one below. Any way to filter those out? 
<img width="792" alt="screen shot 2015-07-17 at 7 59 10 am" src="https://cloud.githubusercontent.com/assets/2094614/8746681/d38cf93e-2c59-11e5-8eb1-4bb27669e9cd.png">

**Comment by ProxiBlue** [comment]
Hi,

If you look at the core code that throws this exception, located in class  Mage_Core_Model_App

public function throwStoreException()
    {
        throw new Mage_Core_Model_Store_Exception('');
    }

you will see that it is actually magento core that throws a message less exception.
This is however a valid exception, resulting from some potential valid error.
It is odd that core throws such a blank error.

If you follow through on the trace, the originator of this exception, on line 829, of teh same class, is:

if (!isset($id)) {
            $this->throwStoreException();
        }
In the context of the calling code is in teh method 'getStore' and  $id = store id

At some point, something in your store is trying to collect the store id, and failing, resulting in teh error.
In my opinion this is a valid exception, and blocking it will hide the fact that something is wrong.

**Comment by netalico** [comment]
As far as I can tell the server is still returning a 200 code and loading the page when those sporadic, unreproducible Mage_Core_Model_Store_Exception errors come up so it would be really nice to be able to exclude them so they don't distract from more serious errors.

**Comment by ProxiBlue** [comment]
Hi,

The problem here is that the exception is a 'Mage_Core_Model_Store_Exception', and additionally it was determined that the exception was thrown because the store (or store id) could not be determined.

Since the store is not known, it would not be possible to load configuration for the store, which in turn means that any admin setting to skip this blank message would not be able to be determined. Trying would just result in an endless exception loop. 

The only way to block this, is by using hard coded check in code. As per my previous answer, I consider this to be  a valid error, and woudl thus not place thus a block, as it may not be desired by the majority of users.

Currently, this issue seems to be isolated to your site, as no other users have complained about the same issue. I have also not been able to reproduce (using magento EE 1.14)

The only solution I can currently offer you is a custom file, which is hard coded to block this exception.
I have placed the updated file in a GIST here: https://gist.github.com/ProxiBlue/5130f80d5f36aab18ba1

Replace the file located here: /app/code/local/Mage/Core/Exception.php with the file.

I still urge you to consider determining the actual reason behind the exception.

**Comment by netalico** [comment]
Thanks for all your help! I'm closing this now.
